vm/src/obj/objlist.rs

```diff
@@ -691,41 +691,51 @@ fn list_new(
 fn quicksort(
     vm: &VirtualMachine,
     keys: &mut [PyObjectRef],
+    indices: &mut [usize],
     values: &mut [PyObjectRef],
 ) -> PyResult<()> {
     let len = values.len();
     if len >= 2 {
-        let pivot = partition(vm, keys, values)?;
-        quicksort(vm, &mut keys[0..pivot], &mut values[0..pivot])?;
-        quicksort(vm, &mut keys[pivot + 1..len], &mut values[pivot + 1..len])?;
+        let pivot = partition(vm, keys, indices, values)?;
+        quicksort(vm, &mut keys[0..pivot], &mut indices[0..pivot], &mut values[0..pivot])?;
+        quicksort(vm, &mut keys[pivot + 1..len], &mut indices[pivot + 1..len], &mut values[pivot + 1..len])?;
     }
     Ok(())
 }
 
 fn partition(
     vm: &VirtualMachine,
     keys: &mut [PyObjectRef],
+    indices: &mut [usize],
     values: &mut [PyObjectRef],
 ) -> PyResult<usize> {
     let len = values.len();
     let pivot = len / 2;
 
     values.swap(pivot, len - 1);
     keys.swap(pivot, len - 1);
+    indices.swap(pivot, len - 1);
 
     let mut store_idx = 0;
     for i in 0..len - 1 {
-        let result = vm._lt(keys[i].clone(), keys[len - 1].clone())?;
-        let boolval = objbool::boolval(vm, result)?;
-        if boolval {
+        let eq = vm._eq(keys[i].clone(), keys[len - 1].clone())?;
```

Comment:
pretty sure CPython doesn't use `__eq__`, so this implementation may fail for classes that do stuff like

    def __eq__(self, other): raise NotImplementedError

CPython's TimSort only uses `__lt__` (and object identity comparisons)